nasa-pictures-stealer/src/main/java/nasapicturesstealer/infrastructure/repository/NasaPicturesRepository.java

```diff
@@ -0,0 +1,72 @@
+package nasapicturesstealer.infrastructure.repository;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import lombok.RequiredArgsConstructor;
+import nasapicturesstealer.domain.entity.NasaCamera;
+import nasapicturesstealer.domain.entity.NasaPicture;
+import nasapicturesstealer.infrastructure.entity.NasaCameraEntity;
+import nasapicturesstealer.infrastructure.entity.NasaPictureEntity;
+import nasapicturesstealer.infrastructure.mapper.NasaCameraMapper;
+import nasapicturesstealer.infrastructure.mapper.NasaPictureMapper;
+import org.springframework.data.util.Pair;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+public class NasaPicturesRepository {
+
+    private final NasaPictureEntityRepository nasaPictureEntityRepository;
+  
+    private final NasaCameraEntityRepository nasaCameraEntityRepository;
+    
+    private final NasaPictureMapper nasaPictureMapper;
+    
+    private final NasaCameraMapper nasaCameraMapper;
+    
+    public List<NasaPicture> saveDomainEntities(List<NasaPicture> pictures) {
+      Map<Integer, Long> nasaIdToPictureId = this.nasaPictureEntityRepository
+        .findAllByNasaIds(pictures.stream().map(NasaPicture::getNasaId).toList())
+        .stream()
+        .collect(Collectors.toMap(Pair::getFirst, Pair::getSecond));
+
+      Map<Integer, Long> nasaIdToCameraId = this.nasaCameraEntityRepository
+        .findAllByNasaIds(pictures.stream().map(NasaPicture::getNasaCamera).map(NasaCamera::getNasaId).toList())
```

Comment:
Another suggestion (But here it is necessary to transfer the method to the appropriate services and also add transactionality): 
```suggestion
package nasapicturesstealer.infrastructure.repository;

import lombok.RequiredArgsConstructor;
import nasapicturesstealer.domain.entity.NasaCamera;
import nasapicturesstealer.domain.entity.NasaPicture;
import nasapicturesstealer.infrastructure.entity.NasaCameraEntity;
import nasapicturesstealer.infrastructure.entity.NasaPictureEntity;
import nasapicturesstealer.infrastructure.mapper.NasaCameraMapper;
import nasapicturesstealer.infrastructure.mapper.NasaPictureMapper;
import org.springframework.stereotype.Component;

import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.function.Function;
import java.util.stream.Collectors;

@Component
@RequiredArgsConstructor
public class NasaPicturesRepository {

    private final NasaPictureEntityRepository nasaPictureEntityRepository;
    private final NasaCameraEntityRepository nasaCameraEntityRepository;
    private final NasaPictureMapper nasaPictureMapper;
    private final NasaCameraMapper nasaCameraMapper;

    public List<NasaPicture> saveDomainEntities(List<NasaPicture> pictures) {
        Map<NasaCamera, List<NasaPicture>> cameraToPicturesMap =
                pictures.stream().collect(Collectors.groupingBy(NasaPicture::getNasaCamera));

        return cameraToPicturesMap.entrySet().stream()
                                  .flatMap(entry -> processCameraPictures(entry.getKey(), entry.getValue()).stream())
                                  .collect(Collectors.toList());
    }

    private List<NasaPicture> processCameraPictures(NasaCamera camera,
                                                    List<NasaPicture> picturesList) {
        Optional<NasaCameraEntity> nasaCameraOpt = nasaCameraEntityRepository.findByNasaId(camera.getNasaId());
        return nasaCameraOpt.map(cameraEntity -> processExistingCamera(picturesList, cameraEntity))
                .orElseGet(() -> processNewCamera(camera, picturesList));
    }

    private List<NasaPicture> processExistingCamera(List<NasaPicture> picturesList,
                                                    NasaCameraEntity cameraEntity) {
        Map<Integer, NasaPictureEntity> storedPictures = getStoredPictures(picturesList);
        List<NasaPictureEntity> newPictures = getNewPictures(picturesList,
                                                             cameraEntity,
                                                             storedPictures);

        return saveAndMapPictures(newPictures, storedPictures.values());
    }

    private List<NasaPicture> processNewCamera(NasaCamera camera,
                                               List<NasaPicture> picturesList) {
        NasaCameraEntity cameraEntity = nasaCameraEntityRepository.save(nasaCameraMapper.toEntity(camera));
        List<NasaPictureEntity> pictureEntities = createPictureEntities(picturesList,
                                                                        cameraEntity);

        return saveAndMapPictures(pictureEntities, List.of());
    }

    private List<NasaPictureEntity> createPictureEntities(List<NasaPicture> picturesList,
                                                          NasaCameraEntity cameraEntity) {
        return picturesList.stream()
                           .map(picture -> new NasaPictureEntity(cameraEntity,
                                                                 picture.getNasaId(),
                                                                 picture.getImgSrc()))
                           .collect(Collectors.toList());
    }

    private Map<Integer, NasaPictureEntity> getStoredPictures(List<NasaPicture> picturesList) {
        return nasaPictureEntityRepository.findAllByNasaIds(getPictureNasaIds(picturesList))
                                          .stream()
                                          .collect(Collectors.toMap(NasaPictureEntity::getNasaId, Function.identity()));
    }

    private List<NasaPictureEntity> getNewPictures(List<NasaPicture> picturesList,
                                                   NasaCameraEntity cameraEntity,
                                                   Map<Integer, NasaPictureEntity> storedPictures) {
        return picturesList.stream()
                           .filter(picture -> !storedPictures.containsKey(picture.getNasaId()))
                           .map(picture -> new NasaPictureEntity(cameraEntity,
                                                                 picture.getNasaId(),
                                                                 picture.getImgSrc()))
                           .collect(Collectors.toList());
    }

    private List<NasaPicture> saveAndMapPictures(List<NasaPictureEntity> newPictures,
                                                 Collection<NasaPictureEntity> persistedPictures) {
        List<NasaPictureEntity> savedPictures = nasaPictureEntityRepository.saveAll(newPictures);
        if(!persistedPictures.isEmpty()){
            savedPictures.addAll(persistedPictures);
        }
        return mapToDomainEntities(savedPictures);
    }

    private List<NasaPicture> mapToDomainEntities(List<NasaPictureEntity> pictureEntities) {
        return pictureEntities.stream().map(nasaPictureMapper::toDomainEntity)
                              .collect(Collectors.toList());
    }

    private static List<Integer> getPictureNasaIds(List<NasaPicture> picturesList) {
        return picturesList.stream()
                           .map(NasaPicture::getNasaId)
                           .toList();
    }

}
```